docsrc/content/type-reader.fsx

```diff
@@ -89,3 +89,40 @@ and
                                         let! name = resolve t
                                         let! value = resolve d
                                         return (name,value) }
+
+
+(**
+Another usage of the Reader monad is an alternative to dependency injection or currying
+in order to pass around dependencies. The below code comes from [F# Online - Josef Starýchfojtů - FSharpPlus - Advanced FP concepts in F#](https://www.youtube.com/watch?v=pxJCHJgG8ws). You can find the presenter on github as [@starychfojtu](https://github.com/starychfojtu).
+
+Why would you want to do this style?
+
+- When you want to pass around a single environment instead of using dependency injection.
+
+Why wouldn't you want to use this style?
+
+- The downside of this style is that it supposes that your environment is relatively immutable. If you have different lifetimes for different implementation classes dependency injection frameworks can be easier to use.
+*)
+open System
+open FSharpPlus
+open FSharpPlus.Data
+
+type IUserRepository =
+    abstract GetUser : email : string -> string
+
+type IShoppingListRepository =
+    abstract Add : shoppingList : string list -> string list
+
+let getUser email =
+    Reader(fun (env : #IUserRepository) -> env.GetUser email)
```

Comment:
What does the '#' mean there?

Reply:
It means that you don't need exactly a `IUserRepository` but a type that implements `IUserRepository`. The name for this is [flexible type](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/flexible-types). Perhaps we should write some text explaining what the example does?